operations/admin_from_mdb.go

```diff
@@ -0,0 +1,282 @@
+					return errors.New("must specify a project to dump data for")
+				}
+				return nil
+			},
+		),
+		Action: func(c *cli.Context) error {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			outfn := c.String(outputFlagName)
+			project := c.String(projectFlagName)
+			dbName := c.String(dbFlagName)
+			url := c.String(urlFlagName)
+
+			var filters map[string]bson.M
+			if project != "" {
+				filters = map[string]bson.M{
+					"versions": bson.M{"identifier": project},
+					"builds":   bson.M{"branch": project},
+					"tasks":    bson.M{"branch": project},
+				}
+			}
+
+			client, err := mongo.NewClient(options.Client().ApplyURI(url))
+			if err != nil {
+				return errors.Wrap(err, "problem creating mongodb client")
+			}
+
+			connCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+			if err = client.Connect(connCtx); err != nil {
+				return errors.Wrap(err, "problem connecting to mongodb")
+			}
+
+			for _, collection := range collections {
+				coll := client.Database(dbName).Collection(collection)
+				size, err := coll.CountDocuments(ctx, struct{}{})
+				if err != nil {
+					return errors.Wrap(err, "problem finding number of source documents")
+				}
+				if size == 0 {
+					return errors.Errorf("cannot write data from collection ('%s') without documents", collection)
+				}
+			}
+
+			if _, err = os.Stat(outfn); !os.IsNotExist(err) {
+				return errors.Errorf("cannot export to %s, file already exists", outfn)
+			}
+
+			var f io.Writer
+			if outfn != "" {
+				f, err = os.Create(outfn)
+				if err != nil {
+					return errors.Wrapf(err, "problem opening file %s", outfn)
+				}
+			} else {
+				f = &bytes.Buffer{}
+			}
+			gw := gzip.NewWriter(f)
+			defer gw.Close()
+			tw := tar.NewWriter(gw)
+			defer tw.Close()
+			var collBuf *bytes.Buffer
+			var filter bson.M
+			var ok bool
+
+			for _, collection := range collections {
+				collBuf = &bytes.Buffer{}
+
+				coll := client.Database(dbName).Collection(collection)
+				if filter, ok = filters[collection]; !ok {
+					filter = bson.M{}
+				}
+				cursor, err := coll.Find(ctx, filter)
+				if err != nil {
+					return errors.Wrap(err, "problem finding documents")
+				}
+
+				count := 0
+
+				for cursor.Next(ctx) {
+					_, err := collBuf.Write(cursor.Current)
+					if err != nil {
+						return errors.Wrap(err, "problem writing document")
+					}
+					count++
+				}
+
+				grip.Info(message.Fields{
+					"count":      count,
+					"collection": collection,
+					"database":   dbName,
+				})
+
+				hdr := &tar.Header{
+					Name: collection,
+					Mode: 0600,
+					Size: int64(collBuf.Len()),
+				}
+				if err := tw.WriteHeader(hdr); err != nil {
+					return errors.Wrap(err, "problem writing tar header")
+				}
+				if _, err := tw.Write(collBuf.Bytes()); err != nil {
+					return errors.Wrap(err, "problem writing buffer to tarball")
+				}
+
+				grip.Error(cursor.Close(ctx))
+			}
+
+			return nil
```

Comment:
this should probably error sometime?

Reply:
It would have returned an error earlier if it did.

Reply:
the error closing the file?